### PR TITLE
Calculate max cluster size correctly for IVF-PQ

### PR DESCRIPTION
**Pull request by tfeher**
This PR fixes the calculation of max cluster size. 

Previous calculation could return much larger than the actual values, which could lead to OOM while allocating temporary buffers while building index for large datasets.

### CODE REVIEW

File: cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh [hunk 1]
```diff
@@ -430,26 +431,21 @@ auto calculate_offsets_and_indices(IdxT n_rows,
...
-  auto max_cluster_size_dev = max_cluster_size_dev_buf.data();
-  update_device(max_cluster_size_dev, &max_cluster_size, 1, stream);
+  auto exec_policy = rmm::exec_policy(stream);
   // Calculate the offsets
   IdxT cumsum = 0;
   update_device(cluster_offsets, &cumsum, 1, stream);
   thrust::inclusive_scan(exec_policy,
                          cluster_sizes,
                          cluster_sizes + n_lists,
                          cluster_offsets + 1,
```
Comment (achirkin) [inline cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh:441]:
```suggestion
```

File: cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh [hunk 2]
```diff
@@ -430,26 +431,21 @@ auto calculate_offsets_and_indices(IdxT n_rows,
...
-                         [max_cluster_size_dev] __device__(IdxT s, uint32_t l) {
-                           atomicMax(max_cluster_size_dev, l);
-                           return s + l;
-                         });
+                         [] __device__(IdxT s, uint32_t l) { return s + l; });
```
Comment (achirkin) [inline cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh:442]:
```suggestion
  thrust::inclusive_scan(
    exec_policy, cluster_sizes, cluster_sizes + n_lists, cluster_offsets + 1, thrust::plus<IdxT>{});
```
I guess, now we can use available functors for this (and force both arguments to be IdxT).

**Review by achirkin — APPROVED** [review]
Thanks for getting to the root of this, Tamas! Looks good.

File: cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh (hunk 2 above)
Comment (tfeher) [reply cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh]: Makes sense, updated

**Review by cjnolet — APPROVED** [review]

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**